### PR TITLE
CXSPA-3072: Rxjs 7 compatible updates

### PR DESCRIPTION
**Pull request by ercultimate**

### CODE REVIEW

File: projects/core/src/auth/user-auth/services/auth-http-header.service.ts [hunk 1]
```diff
@@ -273,6 +273,7 @@ export class AuthHttpHeaderService implements OnDestroy {
           // we want to refresh the access token only when it is old.
           // this is a guard for the case when there are multiple parallel http calls
           if (
+            token &&
```
Comment (plabadie) [inline projects/core/src/auth/user-auth/services/auth-http-header.service.ts]: What is the purpose of adding `token &&` ?

File: projects/storefrontlib/cms-components/myaccount/my-coupons/my-coupons.component.ts [hunk 2]
```diff
@@ -93,21 +93,20 @@ export class MyCouponsComponent implements OnInit, OnDestroy {
...
+          this.subscriptionFail(error);
+        })
+    );
+    this.subscriptions.add(
+      this.couponService
+        .getUnsubscribeCustomerCouponResultError()
+        .subscribe((error) => {
+          this.subscriptionFail(error);
+        })
+    );
```
Comment (plabadie) [inline projects/storefrontlib/cms-components/myaccount/my-coupons/my-coupons.component.ts:109]: Should the observer object be used here?

File: feature-libs/cart/base/core/facade/active-cart.service.ts [hunk 3]
```diff
@@ -378,54 +378,63 @@ export class ActiveCartService implements ActiveCartFacade, OnDestroy {
...
+        filter((cartState) => Boolean(cartState?.success || cartState?.error)),
+        // wait for active cart id to point to code/guid to avoid some work on temp cart entity
+        withLatestFrom(this.activeCartId$),
+        filter(
+          ([cartState, cartId]) => !this.isCartCreating(cartState, cartId)
+        ),
+        map(([cartState]) => cartState.value),
+        filter((cart): cart is Cart => !isEmpty(cart)),
+        take(1)
+      );
```
Comment (plabadie) [inline feature-libs/cart/base/core/facade/active-cart.service.ts:436]: It's a bit hard to see the diff.  Can you help me understand the change? It looks like there are 2 new `pipe().` some new `?.` and `cart is Cart`.  ( all of which I presume must resolve an rxjs 7 issue)

File: projects/core/src/auth/user-auth/services/auth-http-header.service.ts (hunk 1 above)
Comment (ercultimate) [reply projects/core/src/auth/user-auth/services/auth-http-header.service.ts]: discussion results: This results in a logical change, so we will defer this to investigation for 7.0 release.  For now, we will do a type casting.

File: feature-libs/cart/base/core/facade/active-cart.service.ts (hunk 3 above)
Comment (ercultimate) [reply feature-libs/cart/base/core/facade/active-cart.service.ts:436]: Correct, the pipe operator only supports up to 10 arguments before the typing gets too expensive to infer and falls back to `Observable<any>`. To keep the type information available, I broke the pipe arguments down into multiple chained calls with less than 10 arguments.  This is functionally equivalent to all pipe operators being in a single call.

The optional chaining operators are unnecessary and will be removed.

The observable type did not match the return type due to still including `undefined` in the type union.  I adjusted the `filter` function's return type to be a type guard to allow the `filter` to narrow the stream emit type.

File: projects/storefrontlib/cms-components/myaccount/my-coupons/my-coupons.component.ts (hunk 2 above)
Comment (ercultimate) [reply projects/storefrontlib/cms-components/myaccount/my-coupons/my-coupons.component.ts:109]: No, only when using multiple callback functions is the listener object required.  When using a singular "next" callback function, we are not required to use a listener object.

**Review by plabadie — APPROVED** [review]

**Pull request merged**